[REBASE & FF] MdeModulePkg: Create IoMmuLib and add support to - PciBusDxe, PciHostBridgeDxe, NonDiscoverablePciDeviceDxe

eeshanl opened the pull request:
## Description

Reverts PcdRequireIommu change, removes PcdRequireIommu.
Create IoMmuLib
Create IoMmuLibNull
Wrapper library for IoMmu Protocol functions.

Add IoMmuLib support to NonDiscoverablePciDeviceDxe
Modify IoMmu protocol usage and add IoMmuLib support to PciBusDxe and PciHostBridgeDxe

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [x] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

Tested on QEMU SBSA, physical arm device, physical x64 device.

## Integration Instructions

### Platform Integration

The library automatically adapts to platform capabilities. On platforms without IOMMU support, use IoMmuLibNull.
On platforms with IOMMU support, use IoMmuLib.

For integrating with a platform, in the top level DSC, you can do the following for example:

```text
  # Enable IoMmu/Smmu support
  DEFINE REQUIRE_IOMMU             = TRUE

!if $(REQUIRE_IOMMU) == TRUE
  IoMmuLib|MdeModulePkg/Library/IoMmuLib/IoMmuLib.inf
!else
  IoMmuLib|MdeModulePkg/Library/IoMmuLibNull/IoMmuLibNull.inf
!endif
```

eeshanl committed:
Create IoMmuLib

Wrapper library for IoMmu Protocol functions.

eeshanl committed:
Add IoMmu support to NonDiscoverablePciDeviceDxe

Uses IoMmuLib to do IoMmu protocol mappings.

eeshanl commented:
Will add IoMmuLib support for PciIo and PciRootBridgeIo in a separate PR.

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  );
+
+/**
+  Unmap a device address in the Page Table, invalidate the TLB with TLBI operation.
```
TLBI is AARCH64 specific, we should just say invalidate TLB by VA (as I'm assuming it is doing) or don't mention the TLB flush as that is library implementation detail, not relevant to the consumer.

eeshanl replied:
@os-d updated pr: https://github.com/microsoft/mu_silicon_arm_tiano/pull/374 to switch from TLBI ALL to TLBI_S2_IPA InputAddress of Stage 2 SMMU.

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+**/
+EFI_STATUS
+EFIAPI
+IoMmuMap (
```
Do these APIs contain any changes you want to make to IOMMU APIs?

eeshanl replied:
yes, added StreamId and SmmuBase

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+
+  @retval EFI_SUCCESS            The requested pages were freed.
+  @retval EFI_INVALID_PARAMETER  Memory is not a page-aligned address or Pages is invalid.
+  @retval EFI_NOT_FOUND          The requested memory pages were not allocated with AllocatePages().
```
return statuses need updating for this function (looks like copy/paste)

eeshanl replied:
fixed comments throughout

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+                                AllocateMaxAddress or AllocateAddress.
+                                2) MemoryType is in the range
+                                EfiMaxMemoryType..0x6FFFFFFF.
+                                3) Memory is NULL.
```
HostAddress

eeshanl replied:
fixed comments throughout

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+
+  @param [in]  DeviceHandle  The device handle to set attributes for.
+  @param [in]  Mapping       The mapping to set attributes for.
+  @param [in]  IoMmuAccess   The IOMMU access attributes for R/W.
```
Are the only possible attributes (now and forever) RW?

eeshanl replied:
dropped R/W from comment.

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  Set the R/W access attributes for Mapping in the Page Table.
+
+  @param [in]  DeviceHandle  The device handle to set attributes for.
+  @param [in]  Mapping       The mapping to set attributes for.
```
As this is a VOID *, worth mentioning this is the mapping returned by IoMmuMap

eeshanl replied:
yes, done

os-d commented on MdeModulePkg/Library/IoMmuLib/IoMmuLib.inf:
```diff
@@ -0,0 +1,21 @@
+  INF_VERSION                    = 0x00010005
+  BASE_NAME                      = IoMmuLib
+  FILE_GUID                      = 55af180d-a4ee-47fa-84d5-01071643ad35
+  MODULE_TYPE                    = BASE
```
Can't be BASE, uses boot services, which it should also have a LibraryClass dependency on UefiBootServicesTableLib, BaseLib, and DebugLib

eeshanl replied:
what should the type be instead?

os-d replied:
It should be `DXE_DRIVER` because it uses services only available in DXE

eeshanl replied:
done, fixed. thank you.

os-d commented on MdeModulePkg/Library/IoMmuLibNull/IoMmuLibNull.c:
```diff
@@ -0,0 +1,115 @@
+  OUT    VOID                   **Mapping
+  )
+{
+  return EFI_UNSUPPORTED;
```
Should these return EFI_SUCCESS so that a consumer doesn't have to know whether the null instance is being used or not?

eeshanl replied:
In PciIo, and PciRootBridgeIo, the IoMmu map, and unmap call's statuses are imediatly returned, whereas in NonDiscoverable, we continue on with the rest of the function after the IoMmu map or unmap. Due to this difference, we need to be able to distinguish between a NULL implementation to skip the logic essentially, or a real one.

os-d replied:
Are you saying that in the nondiscoverable case we have some logic that we only do if an IoMmu is attached? Or just that there is additional logic that occurs in the driver after?

If it is the latter, it doesn't need to distinguish, right, it can just check if EFI_ERROR, return (or handle appropriately), otherwise continue.

eeshanl replied:
@os-d @cfernald 

https://github.com/microsoft/mu_silicon_intel_tiano/blob/dev/202502/IntelSiliconPkg/Feature/VTd/IntelVTdCoreDxe/IntelVTdCoreDxe.c#L326 turns out a real implementation can return EFI_UNSUPPORTED, so this would no longer work.

I updated the IoMmuLib to have a new function IoMmuIsPresent() . This will return true/false depending on the lib used. And allow the caller to properly use the IoMmu functions within PciRootBridgeIo.

os-d commented on MdeModulePkg/Library/IoMmuLibNull/IoMmuLibNull.c:
```diff
@@ -0,0 +1,115 @@
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/IoMmuLib.h>
+#include <Library/UefiBootServicesTableLib.h>
```
Don't need DebugLib.h or UefiBootServicessTableLib.h

eeshanl replied:
done

os-d commented on MdeModulePkg/Library/IoMmuLibNull/IoMmuLibNull.inf:
```diff
@@ -0,0 +1,14 @@
+
+[Sources]
+  IoMmuLibNull.c
+
```
LibraryClasses dependency on BaseLib

eeshanl replied:
done

os-d commented on MdeModulePkg/MdeModulePkg.dec:
```diff
@@ -35,6 +35,9 @@
   NULL|Include/Library/VariablePolicyLockingExLib.h
+
+  ## @libraryclass IoMmu library that can be used for IoMmu Protocol functions.
+  IoMmuLib|Include/Library/IoMmuLib.h
```
Maybe put in own MU_CHANGE section?

eeshanl replied:
done

os-d commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+/** @file IoMmuLib.h
```
Maybe a small README with the library to indicate why one is chosen over the other, relationship to PCD (or if we are getting rid of PCD), why this is used over protocol directly, etc.

eeshanl replied:
done, added.

os-d commented on MdeModulePkg/Bus/Pci/NonDiscoverablePciDeviceDxe/NonDiscoverablePciDeviceIo.c:
```diff
@@ -1396,6 +1409,60 @@ NonCoherentPciIoMap (
+             DeviceAddress,
+             &MapInfo->IoMmuContext
+             );
+  if (Status != EFI_UNSUPPORTED) {
```
If null lib is changed, then changing this.

eeshanl replied:
per the other comment on unsupported vs success, we would need to keep it like this because of how the iommu functions are called in PciRootBridgeIo vs NonDiscoverable.

os-d replied:
Do we though? Why can't we have the null return Success and just remove the if Status != EFI_UNSUPPORTED? We can just check for an error and fail out, right?

cfernald replied:
I tend to agree with @os-d, If the platform is deciding to use the NULL instance, then it is effectively stubbing out the functionality under the assumption it is not needed and so returning EFI_UNSUPPORTED doesn't really add anything. Especially since this might be a real returned error from other instances.

eeshanl replied:
If you look at the IoMmuMap calls inside RootBridgeIo, it is wrapped around a if (mIommu != NULL) . As a result, when no iommu is present that whole block of logic is skipped. With the NULL IoMmuLibNull, that outter if statement is gone and the function will execute no matter what. We want it to skip the logic in this case and not return the status. So if i have both Libs return SUCCESS, i won't be able to tell when to skip or return the Status.

eeshanl replied:
https://github.com/microsoft/mu_silicon_intel_tiano/blob/dev/202502/IntelSiliconPkg/Feature/VTd/IntelVTdCoreDxe/IntelVTdCoreDxe.c#L326 turns out a real implementation can return EFI_UNSUPPORTED, so this would no longer work.

I updated the IoMmuLib to have a new function IoMmuIsPresent() . This will return true/false depending on the lib used. And allow the caller to properly use the IoMmu functions within PciRootBridgeIo.

cfernald commented on MdeModulePkg/Bus/Pci/NonDiscoverablePciDeviceDxe/NonDiscoverablePciDeviceIo.c:
```diff
@@ -1396,6 +1409,60 @@ NonCoherentPciIoMap (
   }
 
   *Mapping = MapInfo;
```
Will we leak this in even of a lower failure? Because we will set the mapping and not do the free below this point, however we are returning an error so the caller would shouldn't assume its valid.

eeshanl replied:
Fixed to free/unmap on error.

makubacki commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+/**
+  Unmap a device address in the Page Table, invalidate the TLB with TLBI operation.
+
+  @param [in]  This      Pointer to the IOMMU protocol instance.
```
Can drop this parameter.

eeshanl replied:
done

cfernald commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  @param [in, out] NumberOfBytes   On input, the number of bytes to map. On output, the number of bytes mapped.
+  @param [out]     DeviceAddress   The resulting device address.
+  @param [out]     Mapping         A handle to the mapping. Used by IoMmuUnmap to unmap the address and IoMmuSetAttribute to set attributes.
+  @param [in]      StreamId        Optional SMMU stream ID for the mapping.
```
Change SMMU to IOMMU here.

eeshanl replied:
updated comments and renamed to RequesterId and IoMmuBase

eeshanl replied:
Per our conversation in person, we are choosing to drop these extra optional params and will wait to modify this once we make the changes to the IoMmu protocol.

cfernald commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  @param [in, out] NumberOfBytes   On input, the number of bytes to map. On output, the number of bytes mapped.
+  @param [out]     DeviceAddress   The resulting device address.
+  @param [out]     Mapping         A handle to the mapping. Used by IoMmuUnmap to unmap the address and IoMmuSetAttribute to set attributes.
+  @param [in]      StreamId        Optional SMMU stream ID for the mapping.
```
Should detail the expected behavior if the StreamId and/or the SmmuBase is NULL.

eeshanl replied:
done , added additional detail

eeshanl replied:
Per our conversation in person, we are choosing to drop these extra optional params and will wait to modify this once we make the changes to the IoMmu protocol.

cfernald commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  );
+
+/**
+  Allocate a buffer for use with the IOMMU.
```
Could you clarify what "use with the IOMMU" means? i.e. is DMA-able? is <4g? etc..

eeshanl replied:
done

cfernald commented on MdeModulePkg/Include/Library/IoMmuLib.h:
```diff
@@ -0,0 +1,122 @@
+  IN OUT UINTN                  *NumberOfBytes,
+  OUT    EFI_PHYSICAL_ADDRESS   *DeviceAddress,
+  OUT    VOID                   **Mapping,
+  IN     UINTN                  StreamId         OPTIONAL,
```
make pointers for OPTIONAL

eeshanl replied:
done

eeshanl replied:
Per our conversation in person, we are choosing to drop these extra optional params and will wait to modify this once we make the changes to the IoMmu protocol.

os-d commented on MdeModulePkg/Bus/Pci/PciBusDxe/PciBus.c:
```diff
@@ -286,14 +285,6 @@ PciBusDriverBindingStart (
-           &gEdkiiIoMmuProtocolGuid,
-           NULL,
-           (VOID **)&mIoMmuProtocol
-           );
```
mu_change tags in this file (missing in a few other places, too)

eeshanl replied:
Fixed MU_CHANGE tags here and throughout.
Commented out the deleted lines and added MU_CHANGE  tags for those as well.

os-d commented on MdeModulePkg/Bus/Pci/PciHostBridgeDxe/PciRootBridgeIo.c:
```diff
@@ -1666,35 +1657,35 @@ RootBridgeIoAllocateBuffer (
   RootBridge = ROOT_BRIDGE_FROM_THIS (This);
 
-  // MU_CHANGE [BEGIN] - Allow platform to disallow unprotected DMA access when done before IOMMU install.
-  if (FeaturePcdGet (PcdRequireIommu) && (mIoMmu == NULL)) {
```
Are we entirely reverting the PcdRequireIommu change or is it still being used anywhere? If reverting, can we do that as a separate commit (i.e. do the git revert and check this in as rebase & FF) so that we can track that for the next integration?

eeshanl replied:
created an individual revert commit for the PcdRequireIommu change and marked PR for rebase & FF.

eeshanl committed:
Revert "Require IOMMU protocol install before DMA access."

This reverts commit b884f58190904dac7bebd29244efb395421e555d.

eeshanl committed:
Use IoMmuLib in PciBusDxe and PciHostBridgeDxe

- Removes usage of PCD and callbacks to locate IoMmu protocol
- Library wraps all this logic

eeshanl commented on MdeModulePkg/Bus/Pci/PciBusDxe/PciIo.c:
```diff
@@ -1007,30 +1007,62 @@ PciIoMap (
+
+    // MU_CHANGE [BEGIN] - Use IoMmuLib
+    Status = IoMmuSetAttribute (
+               PciIoDevice->Handle,
```
change to PciIoDevice
